lib/axon/compiler.ex

```diff
@@ -129,13 +158,24 @@ defmodule Axon.Compiler do
           params
         end)
 
+      {params, %{parameters: block_parameters, state: block_state}} = normalize_blocks(params)
       params = merge_params!(params, init_params)
 
+      model_state_meta =
+        model_state_meta
+        |> Map.update!(:parameters, &Map.merge(&1, block_parameters))
+        |> Map.update!(:state, &Map.merge(&1, block_state))
```

Comment:
Instead of merging after the fact, pass `model_state_meta.parameters` and `model_state_meta.state` to `normalize_blocks` and already store new key-value directly into them, avoiding this merge operation here.

Quick question: do you need to use the logic in `merge_params!` to also merge the state parameters? What is the difference between the model state data/parameters/frozen_parameters/state fields?

Especially because only the first one is kept as a container, everything else would become part of the Nx.Defn cache key, so picking one over the other matters a lot!